### DISCUSSION TITLE
How to add an indicator and filter in Custom script UI?

### DISCUSSION THREAD

**Discussion by jk-onnet** [netbox-community/netbox#8954] [Q&A]
I am writing a custom script to provision virtual machine.

1) I'm bad in frontend development... May I know how to add a text that appear next to the form field? I wish to know what is the current "cluster total VM" disk space (custom field) when user is selecting the disk size. Is it developing a plugin?

2) How to dynamic filter a list based on the object custom field?
For example:

disk = ChoiceVar(
[10,20,50,100],
label="Disk (GB)",
required=True,
description="CPU")

I want to filter out 10 and 20, if the user select device == "server1".

**Comment by candlerb** [comment]
You could use this one as a starting point: https://github.com/netbox-community/reports/blob/master/scripts/create_vm.py

> May I know how to add a text that appear next to the form field?

There are the 'label' and 'description' attributes on the form field: 'label' appears to the left of the form field, and 'description' appears below it.  What is it you want instead?

> I wish to know what is the current "cluster total VM" disk space (custom field) when user is selecting the disk size

I'm not sure what you're saying here.  Are you saying that the user first selects a Cluster (through an ObjectVar field), and based on which cluster they selected, this changes what options are available for a subsequent field?  If so, that sounds similar to your second question.

However in either case, I'm not aware of a way to do this.  The custom script presents a static form to the user.  Having the form change based on what other options are selected would require some sort of dynamic Javascript updates and/or AJAX to query objects at the back-end.

You can of course do validation within your script, call log_failure if the user choses an invalid selection, and they can hit Back to correct it.  It's not ideal though, and it would be nice if there were a way to perform validation which returns the user directly to the form.

```
from extras.scripts import Script, StringVar, ChoiceVar

class FooBar(Script):
    class Meta:
        name = "FooBar"
        description = "Test validation of fields"

    server_name = StringVar(
        label="Server Name",
        description="This is some text to display next to the field",
        required=True,
    )
    disk = ChoiceVar(
        choices=[("10","10"),("20","20"),("50","50"),("100","100")],
        label="Disk (GB)",
        required=True,
        description="CPU",
    )

    def run(self, data, commit):
        server_name = data["server_name"]
        disk = data["disk"]
        if server_name == "server1" and disk not in ["50", "100"]:
            return self.log_failure("Invalid choice of disk space")
```

**Reply by jk-onnet** [reply]
“ However in either case, I'm not aware of a way to do this. The custom script presents a static form to the user. Having the form change based on what other options are selected would require some sort of dynamic Javascript updates and/or AJAX to query objects at the back-end. “

I think this is what I’m looking for, the main purpose is the user cannot over provisioning the cpu core and disk space before submitting the form instead of submitting the form.
 
if server_name == "server1" and disk not in ["50", "100"]:
            return self.log_failure("Invalid choice of disk space")

This is current my way on doing it but not too interactive. When user click back, all the filled data is gone because I use much of ObjectVar. Is this normal behavior?

Is the plugin revamp release can solve it? Maybe I should learn and wait for 3.2 plugin release on doing javascript updates because I don’t dare to change the code base. Thank you for the reply.

**Reply by candlerb** [reply]
Plugins and custom scripts are distinct features.  I haven't seen any notice of changes to custom scripts in 3.2 (although there might be some).
